### PR TITLE
Add more class declaration formations

### PR DESCRIPTION
**Pull request by jbhoosreddy**
Related Issue: https://github.com/tc39/test262/issues/1343#issuecomment-446225159

This PR attempts to adds additional class declaration formations as per @littledan's test plan

In particular:

   - [x] Types of declarations
       - [x] export default class
       - [x] export class X
       - [x] class declaration at the top level of a script/module
       - [x] class declaration in a block
       - [x] class declaration inside a function at the top level

### CODE REVIEW

**Comment by leobalter** [comment]
Can you push the generated tests too? (in a separate commit)

**Comment by caiolima** [comment]
Also, maybe in another PR, but we are missing covering of `__lookupGetter__` and `__lookupSetter__` on accessors. I think we should create new templates for `private-accessors` case. Right now we are using `private-methods` on those tests.

File: src/class-elements/private-methods/cls-decl-export-default.template [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+esid: prod-MethodDefinition
+---*/
+
+/*** template notes
+ * method should always be #m
+ * the template provides c.ref() for external reference
+ */
+
+function hasProp(obj, name, expected, msg) {
```
Comment (rwaldron) [inline src/class-elements/private-methods/cls-decl-export-default.template]: I'm curious to know why this is here and you're not using `verifyProperty` via `includes: [propertyHelper.js]`

**Comment by leobalter** [comment]
I'm afraid this adds too much without much useful coverage.

It's probably good for tests verifying only the syntactical production, but not for multiplying so many tests. The formats used here are well tested in other places, and maybe these templates should be limited to a minimal production helps leaning the new additions to something less than 7k LOC. WDYT?

cc @gsathya

File: src/class-elements/private-methods/cls-decl-export-default.template (hunk 1 above)
Comment (leobalter) [reply src/class-elements/private-methods/cls-decl-export-default.template]: this is from the other templates we already have and could be improved. hasProp is always used for non existing own properties, and verifyProperty is meant for expected to exist properties.

this could be shortened to `hasNoProp(obj, name, msg)` and doing the following assertions:

```js
function hasNoProp(obj, name, msg) {
  var hasOwnProperty = Object.prototype.hasOwnProperty.call(obj, name);
  assert.sameValue(hasOwnProperty, false, msg);

  var hasProperty = Reflect.has(obj, name);
  assert.sameValue(hasProperty, false, msg);

// or `name in obj` instead of Reflect.has
}
```

Comment (leobalter) [reply src/class-elements/private-methods/cls-decl-export-default.template]: in any case, this should observe any conclusion for https://github.com/tc39/test262/pull/2197#issuecomment-503722953

**Comment by gsathya** [comment]
From an engine perspective, there is difference in how classes are parsed in a module, top level script, inside a block, inside a function (both eager and lazy parsed) -- they trigger different scope analysis paths, so in general these seem ok to me. But I also see @leobalter's concerns here as the spec doesn't really differentiate all these cases as they're more of an "implementation detail" which isn't really under the purview of test262 and these are a lot of tests too. I haven't looked at this in detail but I wonder could just trim down these to test the different scoping rules.

**Comment by littledan** [comment]
Pragmatically, what if we include a few tests for classes in these different environments, without generating all the combinations?

**Comment by leobalter** [comment]
@littledan:

> Pragmatically, what if we include a few tests for classes in these different environments, without generating all the combinations?

This matches my comments:

> It's probably good for tests verifying only the syntactical production

We could extend the templates at the `src/class-elements/syntax/valid` folder for the productions tested here, instead of full cases from `src/class-elements/private-methods`.

File: src/class-elements/private-methods/cls-decl-export-default.template (hunk 1 above)
Comment (rwaldron) [reply src/class-elements/private-methods/cls-decl-export-default.template]: Can we move `hasProp` to `propertyHelper.js` so there isn't many copies of it floating around?

Comment (leobalter) [reply src/class-elements/private-methods/cls-decl-export-default.template]: yes, but this needs to be done in another PR for the existing tests. It should be quick. I'm addressing it tomorrow.

**Comment by littledan** [comment]
Sure, or we could even include a few individual tests (that could also test basic functionality) and not extend any templates.

**Comment by leobalter** [comment]
> Sure, or we could even include a few individual tests (that could also test basic functionality) and not extend any templates.

yes, this would be fine too

**Comment by jbhoosreddy** [comment]
Thanks @gsathya @littledan @leobalter and @rwaldron. So I'll follow @leobalter's suggestion and extend these templates `src/class-elements/syntax/valid`.

If you'd like me to go the other route and not extend templates, I can do that also.

**Comment by jbhoosreddy** [comment]
@leobalter I added those forms in the templates here `/class-elements/syntax/valid` but it still produced 88 files. I'll try to see tomorrow if these tests are worth the maintenance. I may end up writing some non-templated tests instead.

**Comment by jbhoosreddy** [comment]
In today's call @caiolima and I discussed that there might be some value to having these templated tests.
so I'll probably wait for @leobalter 's thoughts on the PR in it's current form.
Happy weekend!

**Comment by jbhoosreddy** [comment]
ping @leobalter for thoughts on this

File: src/class-elements/syntax/valid/cls-decl-elements-valid-syntax-module.template [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+esid: prod-ClassElement
+name: class declaration module
+path: language/statements/class/elements/syntax/valid/module-
+flags: [module]
+features: [class]
+---*/
+
+class C /*{ heritage }*/{
+  /*{ elements }*/
+}
```
Comment (leobalter) [inline src/class-elements/syntax/valid/cls-decl-elements-valid-syntax-module.template]: how is this related to a module? Just the fact it asserts a class inside a module?

It should better to remove this template. There is a template for Script code that can be reused by the runners as modules if necessary. Otherwise this opens a precedent for every syntax we test in this suite.

File: src/class-elements/syntax/valid/cls-decl-elements-valid-syntax-inside-function.template [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+features: [class]
+---*/
+
+function getC() {
+  class C /*{ heritage }*/{
+    /*{ elements }*/
+  }
+  return C;
+}
+var C = getC();
```
Comment (leobalter) [inline src/class-elements/syntax/valid/cls-decl-elements-valid-syntax-inside-function.template]: I believe this is really excessive. A single test to verify the production of a class inside a function would me more than enough.

Also, the `var C = getC();` is going beyond a test for valid syntax.

I recommend removing this.

File: src/class-elements/syntax/valid/cls-decl-elements-valid-syntax-block.template [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+name: class declaration inside block
+path: language/statements/class/elements/syntax/valid/inside-function-
+features: [class]
+---*/
+
+{
+  class C /*{ heritage }*/{
+    /*{ elements }*/
+  }
+}
```
Comment (leobalter) [inline src/class-elements/syntax/valid/cls-decl-elements-valid-syntax-block.template]: a class declaration inside a block is a class declaration and we should limit the tests to itself alone, no expanding to class elements.

I recommend removing this like in src/class-elements/syntax/valid/cls-decl-elements-valid-syntax-inside-function.template

File: src/class-elements/syntax/valid/cls-decl-elements-valid-syntax-export.template [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+esid: prod-ClassElement
+name: class declaration module export
+path: language/statements/class/elements/syntax/valid/export-
+flags: [module]
+features: [class]
+---*/
+
+export class C /*{ heritage }*/{
+  /*{ elements }*/
+}
```
Comment (leobalter) [inline src/class-elements/syntax/valid/cls-decl-elements-valid-syntax-export.template]: I still believe a single test containing a `export ClassDeclaration` is sufficient. This does not observe special syntax for elements in this position.

File: src/class-elements/syntax/valid/cls-decl-elements-valid-syntax-export-default.template [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+esid: prod-ClassElement
+name: class declaration module default export
+path: language/statements/class/elements/syntax/valid/export-default-
+flags: [module]
+features: [class]
+---*/
+
+export default class C /*{ heritage }*/{
+  /*{ elements }*/
+}
```
Comment (leobalter) [inline src/class-elements/syntax/valid/cls-decl-elements-valid-syntax-export-default.template]: Maybe this is the only compelling template in this PR, IMO as the class declaration has an odd variation that is not found elsewhere.

In this case, the variation would be defining a declaration without a BindingIdentifier. This would be a production I believe that is good enough for checking syntax.

In this case, I would just change this template code to be like:

```
export default class /*{ heritage }*/{
  /*{ elements }*/
}
```

**Comment by anba** [comment]
I've removed the tests again as part of #2097 (https://github.com/tc39/test262/pull/2097/commits/64b6b844ebfe1c82645d0ff7f325d4453f338cad), because they were violating <https://github.com/tc39/test262/blob/master/src/class-elements/conventions.md>, which rendered them useless.
